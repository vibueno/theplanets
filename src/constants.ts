const sections = {
  overview: {
    id: 'overview',
    number: '01',
    btnCaption: 'overview',
    btnCaptionExtended: 'overview'
  },
  structure: {
    id: 'structure',
    number: '02',
    btnCaption: 'structure',
    btnCaptionExtended: 'internal structure'
  },
  geology: {
    id: 'geology',
    number: '03',
    btnCaption: 'geology',
    btnCaptionExtended: 'surface geology'
  }
};

const planetNames = {
  mercury: 'mercury',
  venus: 'venus',
  earth: 'earth',
  mars: 'mars',
  jupiter: 'jupiter',
  saturn: 'saturn',
  uranus: 'uranus',
  neptune: 'neptune'
};

const planetKeyData = {
  rotation: {
    id: 'rotation',
    title: 'rotation time'
  },
  revolution: {
    id: 'revolution',
    title: 'rotation time'
  },
  radius: {
    id: 'radius',
    title: 'radius'
  },
  temperature: {
    id: 'temperature',
    title: 'average temp.'
  }
};

export { sections, planetNames, planetKeyData };
