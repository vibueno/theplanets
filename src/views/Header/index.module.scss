@use 'SRC/variables' as *;
@use 'SRC/mixins.scss' as mixins;

nav
{
  position: sticky;
  z-index: 2;
  top: 0;

  background: $color-blue-dark url('ASSETS/img/background-stars.svg');

  .appTitle
  {
    font: 500 28px Antonio, Arial, Helvetica;

    display: grid;
    justify-content: left;

    margin: .9rem 0 .9rem 1.25rem;

    letter-spacing: -1.05px;
    text-transform: uppercase;

    grid-row: 1 / 2;
    grid-template-columns: repeat(2, auto);
  }

  .hamburger
  {
    position: fixed;
    z-index: 2;
    top: 1.55rem;
    right: 1.25rem;

    cursor: pointer;
  }

  &[data-menu-open='true']
  {
    grid-template-rows: 4rem auto;
  }
}

@media only screen and (max-width: 767px)
{
  nav
  {
    .appTitle
    {
      margin: 0;
      padding: .9rem 0 .9rem 1.25rem;

      border-bottom: $border;
    }
  }
}

@media only screen and (min-width: 768px)
{
  nav
  {
    border-bottom: $border;

    .hamburger
    {
      display: none;
    }

    .menu
    {
      justify-content: center;
    }

    .appTitle
    {
      justify-content: center;

      margin: .75em 0;

      grid-row: 1 / 2;
    }
  }
}

@media only screen and (min-width: 1200px)
{
  nav
  {
    display: flex;
    align-items: stretch;

    padding: 0 2rem;

    grid-column: 1 / 3;

    .appTitle
    {
      align-items: center;
      flex: 1 1 30%;
      justify-content: flex-start;

      margin: 0;
    }

    .menu
    {
      align-items: stretch;
      flex: 1 1 70%;
      justify-content: flex-end;

      margin: 0;
    }

    .planetSectionsMenu
    {
      display: none;
    }
  }
}
