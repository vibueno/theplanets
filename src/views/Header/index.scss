@import 'SRC/variables.scss';

nav {
  grid-row: 1;
  display: grid;
  grid-template-rows: repeat(3, auto);
  align-items: stretch;
  text-transform: uppercase;

  .app-title {
    margin: 0.9rem 0 0.9rem 1.25rem;
    grid-row: 1 / 2;
    display: grid;
    grid-template-columns: repeat(2, auto);
    justify-content: left;
    font: 500 28px Antonio, Arial, Helvetica;
    letter-spacing: -1.05px;
  }

  .hamburger {
    position: fixed;
    top: 1.55rem;
    right: 1.25rem;
    cursor: pointer;
  }

  .menu {
    display: fixed;
  }

  &[data-menu-open='true'] {
    grid-template-rows: repeat(2, auto);
    grid-template-columns: repeat(2, auto);
  }

  .planet-sections-menu {
    grid-row: 2 / 3;
    grid-column: 1 / 1;
    display: grid;
    grid-template-columns: repeat(3, auto);
    border-bottom: $border;
    border-top: $border;

    .btn {
      justify-content: center;
      padding: 0.9rem 0 0.75rem 0;
      background-color: transparent;
    }

    #overview {
      grid-column: 1 / 2;
    }

    #structure {
      grid-column: 2 / 3;
    }

    #geology {
      grid-column: 3 / 4;
    }
  }
}

@media only screen and (max-width: 767px) {
  nav {
    &[data-menu-open='false'] {
      .menu {
        display: none;
      }
    }
  }
}

@media only screen and (min-width: 768px) {
  nav {
    .planet-sections-menu {
      display: none;
    }

    border-bottom: $border;

    .hamburger {
      display: none;
    }

    .menu {
      display: fixed;
      justify-content: center;
    }

    .app-title {
      grid-row: 1 / 2;
      justify-content: center;
      margin: 0.75em 0;
    }
  }
}
