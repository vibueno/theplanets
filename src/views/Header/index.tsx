import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { setIsMenuOpen } from 'REDUX/appStateSlice';
import { selectors } from 'REDUX/appStateSlice';

import PlanetSections from 'COMPONENTS/planet/PlanetSections';
import Menu from 'COMPONENTS/Menu';
import { APP_TITLE } from '../../constants';

import styles from './index.module.scss';

type HeaderProps = {
  menuClickHandler: (
    e: React.MouseEvent<HTMLLIElement>,
    ref: React.RefObject<HTMLLIElement>
  ) => void;
  sectionMenuClickHandler: (
    e: React.MouseEvent<HTMLButtonElement>,
    ref: React.RefObject<HTMLButtonElement>
  ) => void;
};

const Header = ({ menuClickHandler, sectionMenuClickHandler }: HeaderProps) => {
  const dispatch = useDispatch();

  const currentPlanetKey = useSelector(selectors.getCurrentPlanetKey);
  const currentSectionKey = useSelector(selectors.getCurrentSectionKey);
  const isMenuOpen = useSelector(selectors.getIsMenuOpen);
  const isTransitionDone = useSelector(selectors.getIsTransitionDone);

  const hamburgerClickHandler = () => {
    if (isTransitionDone) {
      dispatch(setIsMenuOpen(!isMenuOpen));
    }
  };

  return (
    <nav data-menu-open={isMenuOpen} data-is-transition-done={isTransitionDone}>
      <div className={styles.appTitle}>{APP_TITLE}</div>
      <img
        src={require(`ASSETS/img/icon-hamburger.svg`)}
        className={styles.hamburger}
        onClick={hamburgerClickHandler}
      />

      <Menu
        isInsideHeader
        currentPlanetKey={currentPlanetKey}
        clickHandler={menuClickHandler}
      />

      <PlanetSections
        currentPlanetKey={currentPlanetKey}
        currentSectionKey={currentSectionKey}
        clickHandler={sectionMenuClickHandler}
        isMenu={true}
      />
    </nav>
  );
};

export default Header;
