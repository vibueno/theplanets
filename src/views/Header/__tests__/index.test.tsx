import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import * as renderer from 'react-test-renderer';
import Header from '../';
import { APP_TITLE, PLANETS, SECTIONS } from 'SRC/constants';

describe('Header', function () {
  const initialState = {
    appState: {
      isMenuOpen: false,
      isTransitionDone: true,
      currentPlanetKey: PLANETS.MERCURY.KEY,
      currentSectionKey: SECTIONS.OVERVIEW.KEY
    }
  };

  const mockStore = configureStore();
  let store;

  it('Header renders correctly', () => {
    store = mockStore(initialState);
    const props = {
      menuClickHandler: () => {
        return;
      },
      hamburgerClickHandler: () => {
        return;
      },
      sectionMenuClickHandler: () => {
        return;
      }
    };
    const tree = renderer
      .create(
        <Provider store={store}>
          <Header {...props} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
