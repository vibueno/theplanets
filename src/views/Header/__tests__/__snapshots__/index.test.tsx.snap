// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Header renders correctly 1`] = `
<nav
  data-is-transition-done={true}
  data-menu-open={false}
>
  <div>
    the planets
  </div>
  <img
    onClick={[Function]}
    src=""
  />
  <ul
    data-is-inside-header={true}
  >
    <li
      className="mercury"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={true}
      id="MERCURY"
      onClick={[Function]}
    >
      <span />
      <span>
        mercury
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="venus"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="VENUS"
      onClick={[Function]}
    >
      <span />
      <span>
        venus
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="earth"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="EARTH"
      onClick={[Function]}
    >
      <span />
      <span>
        earth
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="mars"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="MARS"
      onClick={[Function]}
    >
      <span />
      <span>
        mars
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="jupiter"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="JUPITER"
      onClick={[Function]}
    >
      <span />
      <span>
        jupiter
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="saturn"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="SATURN"
      onClick={[Function]}
    >
      <span />
      <span>
        saturn
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="uranus"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="URANUS"
      onClick={[Function]}
    >
      <span />
      <span>
        uranus
      </span>
      <img
        src=""
      />
    </li>
    <li
      className="neptune"
      data-is-inside-header={true}
      data-menu-open={false}
      data-selected={false}
      id="NEPTUNE"
      onClick={[Function]}
    >
      <span />
      <span>
        neptune
      </span>
      <img
        src=""
      />
    </li>
  </ul>
  <div
    className="undefined"
    data-is-transition-done={true}
    data-menu-open={false}
  >
    <button
      className="undefined undefined"
      data-selected={true}
      id="OVERVIEW"
      onClick={[Function]}
    >
      <span>
        01
      </span>
      <span>
        overview
      </span>
      <span>
        overview
      </span>
    </button>
    <button
      className="undefined undefined"
      data-selected={false}
      id="STRUCTURE"
      onClick={[Function]}
    >
      <span>
        02
      </span>
      <span>
        structure
      </span>
      <span>
        internal structure
      </span>
    </button>
    <button
      className="undefined undefined"
      data-selected={false}
      id="GEOLOGY"
      onClick={[Function]}
    >
      <span>
        03
      </span>
      <span>
        geology
      </span>
      <span>
        surface geology
      </span>
    </button>
  </div>
</nav>
`;
